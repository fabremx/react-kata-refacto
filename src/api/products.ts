import { Product } from "../models";

export const getProducts = async (): Promise<Product[]> => {
  const response = await fetch(`http://localhost:3000/api/products.json`);
  const { products }: { products: Product[] } = await response.json();
  return products;
};
