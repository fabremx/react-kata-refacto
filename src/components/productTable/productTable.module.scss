@import "../../assets/styles/colors.scss";

.cart {
  width: 100%;
  margin: 0 15px;

  table {
    width: 100%;
    max-width: 100%;
    margin-bottom: 1rem;
    background-color: transparent;
    border-collapse: collapse;
  }

  thead {
    background-color: rgba(0, 0, 0, 0.03);
    border: 1px solid rgba(0, 0, 0, 0.125);
    font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto,
      "Helvetica Neue", Arial, sans-serif, "Apple Color Emoji", "Segoe UI Emoji",
      "Segoe UI Symbol";
    height: 40px;
    text-align: left;

    th {
      font-size: 15px;
      padding-left: 10px;
    }
  }
}
