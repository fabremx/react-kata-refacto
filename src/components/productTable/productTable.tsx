import { ReactElement } from "react";
import { Product } from "../../models";
import { ProductRow } from "../productRow";
import styles from "./productTable.module.scss";

interface Props {
  products: Product[];
  updateProduct: (products: Product[]) => void;
}

export function ProductTable({ products, updateProduct }: Props): ReactElement {
  const removeProduct = (idToRemove: number) => {
    const updatedProducts: Product[] = products.filter(
      (product: Product) => product.id !== idToRemove
    );
    updateProduct(updatedProducts);
  };

  return (
    <div className={styles.cart}>
      <table>
        <thead>
          <tr>
            <th>Image</th>
            <th>Product Name</th>
            <th>Price</th>
            <th>Quantity</th>
            <th>Total</th>
            <th>Remove</th>
          </tr>
        </thead>

        <tbody>
          {products.length === 0 && (
            <tr>
              <td
                colSpan={6}
                className={styles.noProducts}
                data-testid="noProducts"
              >
                No products in cart
              </td>
            </tr>
          )}

          {products.map((product) => (
            <ProductRow
              product={product}
              removeProduct={removeProduct}
              key={product.id}
            />
          ))}
        </tbody>
      </table>
    </div>
  );
}
