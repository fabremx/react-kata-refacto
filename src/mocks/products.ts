import { Product } from "../models";

const BASE_PRODUCT = {
  price: 10,
  image: "image 1",
  quantity: 2,
};

export const NEW_PRODUCT: Product = {
  ...BASE_PRODUCT,
  id: 1,
  title: "New Product",
  isNew: true,
  isSoonEnding: false,
};
export const ENDED_PRODUCT: Product = {
  ...BASE_PRODUCT,
  id: 2,
  title: "Ended Product",
  isNew: false,
  isSoonEnding: true,
};
export const COMMON_PRODUCT: Product = {
  ...BASE_PRODUCT,
  id: 3,
  title: "Common Product",
  isNew: false,
  isSoonEnding: false,
};
