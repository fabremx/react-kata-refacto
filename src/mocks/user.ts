import { User } from "../models";

const BASE_USER = {
  id: 1,
  firstname: "firstname",
  lastname: "lastname",
};
export const MOCKED_USER = {
  ...BASE_USER,
  isVIP: false,
  coupon: 0,
};
export const USER_VIP_WITH_COUPON: User = {
  ...BASE_USER,
  isVIP: true,
  coupon: 10,
};
export const USER_VIP_WITHOUT_COUPON: User = {
  ...BASE_USER,
  isVIP: true,
  coupon: 0,
};

export const USER_NOT_VIP_WITH_COUPON: User = {
  ...BASE_USER,
  isVIP: false,
  coupon: 10,
};
export const USER_NOT_VIP_WITHOUT_COUPON: User = {
  ...BASE_USER,
  isVIP: false,
  coupon: 0,
};
