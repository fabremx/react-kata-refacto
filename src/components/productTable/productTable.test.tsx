import { render } from "@testing-library/react";
import { COMMON_PRODUCT } from "../../mocks/products";
import { ProductTable } from "./productTable";

jest.mock("../productRow", () => ({
  ProductRow: () => (
    <tr>
      <td>Product Mock</td>
    </tr>
  ),
}));

describe("ProductPrice", () => {
  it("should render correctly the component when products is empty", () => {
    // When
    const { asFragment } = render(
      <ProductTable products={[]} updateProduct={jest.fn()} />
    );
    // Then
    expect(asFragment()).toMatchSnapshot();
  });

  it("should render correctly the component when products is NOT empty", () => {
    // When
    const { asFragment } = render(
      <ProductTable products={[COMMON_PRODUCT]} updateProduct={jest.fn()} />
    );
    // Then
    expect(asFragment()).toMatchSnapshot();
  });
});
