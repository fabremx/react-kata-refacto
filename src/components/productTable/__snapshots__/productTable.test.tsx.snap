// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductPrice should render correctly the component when products is NOT empty 1`] = `
<DocumentFragment>
  <div
    class="cart"
  >
    <table>
      <thead>
        <tr>
          <th>
            Image
          </th>
          <th>
            Product Name
          </th>
          <th>
            Price
          </th>
          <th>
            Quantity
          </th>
          <th>
            Total
          </th>
          <th>
            Remove
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            Product Mock
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</DocumentFragment>
`;

exports[`ProductPrice should render correctly the component when products is empty 1`] = `
<DocumentFragment>
  <div
    class="cart"
  >
    <table>
      <thead>
        <tr>
          <th>
            Image
          </th>
          <th>
            Product Name
          </th>
          <th>
            Price
          </th>
          <th>
            Quantity
          </th>
          <th>
            Total
          </th>
          <th>
            Remove
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td
            class="noProducts"
            colspan="6"
            data-testid="noProducts"
          >
            No products in cart
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</DocumentFragment>
`;
